### PR TITLE
Fixes #33033 - shell_escape template macro

### PR DESCRIPTION
**Pull request by stejskalleos**
New template macro `shell_escape` using
Ruby's `Shellwords.shellescape` for safe shell strings.

Use `shell_escape` for user input variables
in registration templates (`global` & `host_init`)

**Issue**
Content host registration using command generated from global registration form fails if activation key name contains ' (single quote) in it.

### CODE REVIEW

**Comment by stejskalleos** [comment]
@ares

File: lib/foreman/renderer/scope/macros/helpers.rb [hunk 1]
```diff
@@ -90,6 +90,14 @@ def format_time(time, format: '%Y-%-m-%-d %k:%M:%S %z', zone: Time.zone)
 
             time_to_format.in_time_zone(zone).strftime(format)
           end
+
+          apipie :method, "Escape string to safely use it with a shell" do
+            required :string, String, desc: 'String to espace'
```
Comment (ofedoren) [inline lib/foreman/renderer/scope/macros/helpers.rb]: Typo: `escape`

File: lib/foreman/renderer/scope/macros/helpers.rb [hunk 2]
```diff
@@ -90,6 +90,14 @@ def format_time(time, format: '%Y-%-m-%-d %k:%M:%S %z', zone: Time.zone)
...
+            returns String
```
Comment (ofedoren) [inline lib/foreman/renderer/scope/macros/helpers.rb:96]: Maybe it's worth to add an example?

**Review by ofedoren — COMMENTED** [review]
Just a few nits from me..

**Comment by ezr-ondrej** [comment]
@ofedoren please assign yourself if you are planning to finish the review. Don't if you want someone else to take a look :)

File: app/views/unattended/provisioning_templates/registration/global_registration.erb [hunk 3]
```diff
@@ -85,11 +85,11 @@ register_host() {
 <%= "       --data 'host[location_id]=#{@location.id}' \\\n" if @location -%>
 <%= "       --data 'host[hostgroup_id]=#{@hostgroup.id}' \\\n" if @hostgroup -%>
 <%= "       --data 'host[operatingsystem_id]=#{@operatingsystem.id}' \\\n" if @operatingsystem -%>
-<%= "       --data 'host[interfaces_attributes][0][identifier]=#{@remote_execution_interface}' \\\n" if @remote_execution_interface.present? -%>
+<%= "       --data \"host[interfaces_attributes][0][identifier]=#{@remote_execution_interface}\" \\\n" if @remote_execution_interface.present? -%>
```
Comment (ares) [inline app/views/unattended/provisioning_templates/registration/global_registration.erb]: while the quotes are ok, it's still a user input and needs to be escaped with shell_escape which will also escape spaces

I'd say use shell_escape here and remove the quotes again

File: app/views/unattended/provisioning_templates/registration/global_registration.erb [hunk 4]
```diff
@@ -85,11 +85,11 @@ register_host() {
...
 <%= "       --data 'setup_insights=#{@setup_insights}' \\\n" unless @setup_insights.nil? -%>
 <%= "       --data 'setup_remote_execution=#{@setup_remote_execution}' \\\n" unless @setup_remote_execution.nil? -%>
-<%= "       --data 'remote_execution_interface=#{@remote_execution_interface}' \\\n" if @remote_execution_interface.present? -%>
-<%= "       --data 'packages=#{@packages}' \\\n" if @packages.present? -%>
+<%= "       --data \"remote_execution_interface=#{@remote_execution_interface}\" \\\n" if @remote_execution_interface.present? -%>
```
Comment (ares) [inline app/views/unattended/provisioning_templates/registration/global_registration.erb]: same here

File: app/views/unattended/provisioning_templates/registration/global_registration.erb [hunk 5]
```diff
@@ -85,11 +85,11 @@ register_host() {
...
+<%= "       --data \"packages=#{@packages}\" \\\n" if @packages.present? -%>
```
Comment (ares) [inline app/views/unattended/provisioning_templates/registration/global_registration.erb]: packages is also a user input, so shell_escape is necessary

File: app/views/unattended/provisioning_templates/registration/global_registration.erb [hunk 6]
```diff
@@ -111,8 +111,8 @@ if [ x$ID = xrhel ] || [ x$ID = xcentos ] || [ x$ID = xol ]; then
 <%= "          --data 'host[lifecycle_environment_id]=#{@lifecycle_environment_id}' \\\n" if @lifecycle_environment_id.present? -%>
 <%= "          --data 'setup_insights=#{@setup_insights}' \\\n" unless @setup_insights.nil? -%>
 <%= "          --data 'setup_remote_execution=#{@setup_remote_execution}' \\\n" unless @setup_remote_execution.nil? -%>
-<%= "          --data 'remote_execution_interface=#{@remote_execution_interface}' \\\n" if @remote_execution_interface.present? -%>
-<%= "          --data 'packages=#{@packages}' \\\n" if @packages.present? -%>
+<%= "          --data \"remote_execution_interface=#{@remote_execution_interface}\" \\\n" if @remote_execution_interface.present? -%>
```
Comment (ares) [inline app/views/unattended/provisioning_templates/registration/global_registration.erb]: also here

File: app/views/unattended/provisioning_templates/registration/global_registration.erb [hunk 7]
```diff
@@ -111,8 +111,8 @@ if [ x$ID = xrhel ] || [ x$ID = xcentos ] || [ x$ID = xol ]; then
...
+<%= "          --data \"packages=#{@packages}\" \\\n" if @packages.present? -%>
```
Comment (ares) [inline app/views/unattended/provisioning_templates/registration/global_registration.erb]: and here

**Review by ares — COMMENTED** [review]
I don't want to "request changes" to not block the PR from merging, I think we need to escape more things but it's a minor change, overall :+1: on the solution

File: lib/foreman/renderer/scope/macros/helpers.rb [hunk 8]
```diff
@@ -90,6 +90,15 @@ def format_time(time, format: '%Y-%-m-%-d %k:%M:%S %z', zone: Time.zone)
 
             time_to_format.in_time_zone(zone).strftime(format)
           end
+
+          apipie :method, "Escape string to safely use it with a shell" do
+            required :string, String, desc: 'String to escape'
+            returns String
+            example "shell_escape('It's escaped!') #=> 'It\\'s\\ escaped\\!'"
```
Comment (adamruzicka) [inline lib/foreman/renderer/scope/macros/helpers.rb]: Now this example doesn't really work, does it?

File: app/views/unattended/provisioning_templates/host_init_config/host_init_config_default.erb [hunk 9]
```diff
@@ -47,7 +47,7 @@ trap 'exit_and_cancel_build' ERR
 
 <% end -%>
 
-<%= install_packages(host_param('host_packages')) -%>
+<%= install_packages(shell_escape(host_param('host_packages'))) -%>
```
Comment (adamruzicka) [inline app/views/unattended/provisioning_templates/host_init_config/host_init_config_default.erb]: Looking at the definition, it expects a space delimited list of package names. Won't this break if we escape the it?

**Comment by stejskalleos** [comment]
Rebased & updated: New example and removed escape of `host_param('host_packages')` for `install_packages`

**Review by adamruzicka — APPROVED** [review]
Nothing more to add, lgtm

**Comment by adamruzicka** [comment]
Thank you @stejskalleos !

**Comment by ezr-ondrej** [comment]
CP 2.5: #8707

**Pull request merged**